tensorflow/contrib/io/kernels/zmq_op.cc

```diff
@@ -0,0 +1,108 @@
+    OP_REQUIRES_OK(context, context->GetAttr("address", &address_));
+    OP_REQUIRES_OK(context, context->GetAttr("timeout", &timeout_));
+    // Create a context and connect a REQ socket
+    context_ = zmq_ctx_new();
+    OP_REQUIRES(context, context_ != nullptr,
+                errors::Internal("Failed to initialize context."));
+    socket_ = zmq_socket(context_, ZMQ_REQ);
+    OP_REQUIRES(context, socket_ != nullptr,
+                errors::Internal("Failed to initialize socket."));
+    OP_REQUIRES(context,
+                zmq_setsockopt(socket_, ZMQ_RCVTIMEO,
+                               &timeout_, sizeof(timeout_)) == 0,
+                errors::Internal("Failed to set timeout."));
+    OP_REQUIRES(context, zmq_connect(socket_, &address_[0]) == 0,
+                errors::Internal("Failed to connect to ", address_, "."));
+  }
+
+  ~PollZmqOp() {
+    if (socket_ != nullptr) {
+      zmq_close(socket_);
+    }
+    if (context_ != nullptr) {
+      zmq_ctx_destroy(context_);
+    }
+  }
+
+  using AsyncOpKernel::AsyncOpKernel;
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) override {
+    const Tensor* input;
+    OP_REQUIRES_OK(context, context->input("request", &input));
+    OP_REQUIRES(context, TensorShapeUtils::IsScalar(input->shape()),
+                errors::InvalidArgument(
+                    "Input message tensor must be scalar, but had shape: ",
+                    input->shape().DebugString()));
+
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context, context->allocate_output("reply",
+                                                      TensorShape({}), &output));
+
+    // Get the message and response as string scalars
+    auto request = input->scalar<string>()();
+    auto &reply = output->scalar<string>()();
+
+    // Lock for exclusive access (destructor releases)
+    mutex_lock lock(mu_);
+
+    // Prepare the message
+    zmq_msg_t request_msg;
+    OP_REQUIRES(context, zmq_msg_init_size(&request_msg, request.size()) == 0,
+                errors::Internal("Failed to initialize request."));
+    memmove(zmq_msg_data(&request_msg), &request[0], request.size());
+    // Send the message (this also takes care of clean up)
+    OP_REQUIRES(context, zmq_msg_send(&request_msg, socket_, 0) == request.size(),
+                errors::Internal("Failed to send reply."));
+
+    // Get the response
+    zmq_msg_t reply_msg;
+    OP_REQUIRES(context, zmq_msg_init(&reply_msg) == 0,
+                errors::Internal("Failed to initialize reply."));
+    OP_REQUIRES(context, zmq_msg_recv(&reply_msg, socket_, 0) != -1,
+                errors::Internal("Failed to receive reply within ", timeout_, "ms."));
+    // Copy the data and clean up
+    reply.resize(zmq_msg_size(&reply_msg));
+    memmove(&reply[0], zmq_msg_data(&reply_msg), zmq_msg_size(&reply_msg));
+    OP_REQUIRES(context, zmq_msg_close(&reply_msg) == 0,
+                errors::Internal("Failed to close reply message."));
+    done();
```

Comment:
It might be fine to delay refactoring this until the constructor/accessor is in place (since e.g. a resource for the ZMQ socket would be the logical owner for a background thread), but please either switch back to a synchronous kernel or fix the `OP_REQUIRES()` invocations.